flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableInputFormat.java

```diff
@@ -0,0 +1,327 @@
+		List<HiveTableInputSplit> hiSplit = new ArrayList<>();
+		int splitNum = 0;
+		for (HiveTablePartition partition : partitions) {
+			StorageDescriptor sd = partition.getStorageDescriptor();
+			InputFormat format;
+			try {
+				format = (InputFormat)
+					Class.forName(sd.getInputFormat(), true, Thread.currentThread().getContextClassLoader()).newInstance();
+			} catch (Exception e) {
+				throw new RuntimeException("Unable to instantiate the hadoop input format", e);
+			}
+			ReflectionUtils.setConf(format, jobConf);
+			jobConf.set(INPUT_DIR, sd.getLocation());
+			//TODO: we should consider how to calculate the splits according to minNumSplits in the future.
+			org.apache.hadoop.mapred.InputSplit[] splitArray = format.getSplits(jobConf, minNumSplits);
+			for (int i = 0; i < splitArray.length; i++) {
+				hiSplit.add(new HiveTableInputSplit(splitNum++, splitArray[i], jobConf, partition));
+			}
+		}
+
+		return hiSplit.toArray(new HiveTableInputSplit[hiSplit.size()]);
+	}
+
+	@Override
+	public void configure(org.apache.flink.configuration.Configuration parameters) {
+
+	}
+
+	// Todo: refactor code to get statistics from hms or partition.
+	@Override
+	public BaseStatistics getStatistics(BaseStatistics cachedStats) throws IOException {
+		// only gather base statistics for FileInputFormats
+		if (!(mapredInputFormat instanceof FileInputFormat)) {
+			return null;
+		}
+
+		final org.apache.flink.api.common.io.FileInputFormat.FileBaseStatistics cachedFileStats =
+				(cachedStats instanceof org.apache.flink.api.common.io.FileInputFormat.FileBaseStatistics) ?
+				(org.apache.flink.api.common.io.FileInputFormat.FileBaseStatistics) cachedStats : null;
+
+		try {
+			final org.apache.hadoop.fs.Path[] paths = FileInputFormat.getInputPaths(this.jobConf);
+
+			return HiveTableUtil.getFileStats(cachedFileStats, paths, new ArrayList<FileStatus>(1));
+		} catch (IOException ioex) {
+			if (logger.isWarnEnabled()) {
+				logger.warn("Could not determine statistics due to an io error: "
+							+ ioex.getMessage());
+			}
+		} catch (Throwable t) {
+			if (logger.isErrorEnabled()) {
+				logger.error("Unexpected problem while getting the file statistics: " + t.getMessage(), t);
+			}
+		}
+
+		// no statistics available
+		return null;
+	}
+
+	@Override
+	public InputSplitAssigner getInputSplitAssigner(HiveTableInputSplit[] inputSplits) {
+		return new LocatableInputSplitAssigner(inputSplits);
+	}
+
+	@Override
+	public boolean reachedEnd() throws IOException {
+		if (!fetched) {
+			fetchNext();
+		}
+		return !hasNext;
+	}
+
+	@Override
+	public void close() throws IOException {
+		if (this.recordReader != null) {
+			this.recordReader.close();
+		}
+	}
+
+	protected void fetchNext() throws IOException {
+		hasNext = this.recordReader.next(key, value);
+		fetched = true;
+	}
+
+	@Override
+	public BaseRow nextRecord(BaseRow ignore) throws IOException {
+		if (!this.fetched) {
+			fetchNext();
+		}
+		if (!this.hasNext) {
+			return null;
```

Comment:
why return `null` when there's no next record

Reply:
When there is no new row, return null which will not be collected forwarding to downstream